File: internal/types/costs.go
package types

import (
	"time"

	costexplorertypes "github.com/aws/aws-sdk-go-v2/service/costexplorer/types"
)

type CostQueryTimePeriod struct {
	Start string `json:"start"`
	End   string `json:"end"`
}

type CostQueryInfo struct {
	TimePeriod      CostQueryTimePeriod `json:"time_period"`
	Granularity     string              `json:"granularity"`
	Services        []string            `json:"services"`
	Regions         []string            `json:"regions"`
	GroupBy         []string            `json:"group_by"`
	Metrics         []string            `json:"metrics"`
	Tags            map[string][]string `json:"tags,omitempty"`
	AWSCLICommand   string              `json:"aws_cli_command"`
	ConsoleURL      string              `json:"console_url"`
	AggregationNote string              `json:"aggregation_note"`
}

type CostInformation struct {
	CostMetadata CostMetadata                     `json:"metadata"`
	CostResults  []costexplorertypes.ResultByTime `json:"results"`
	QueryInfo    CostQueryInfo                    `json:"query_info"`
}

type CostMetadata struct {
	StartDate   time.Time           `json:"start_date"`
	EndDate     time.Time           `json:"end_date"`
	Granularity string              `json:"granularity"`
	Tags        map[string][]string `json:"tags"`
	Services    []string            `json:"services"`
}

File: cmd/ui/frontend/src/types/api/costs.ts
import type { ApiMetadata } from './common'

/**
 * Cost value structure - can contain multiple cost types
 */
export interface CostValues {
  unblended_cost?: number
  blended_cost?: number
  amortized_cost?: number
  net_amortized_cost?: number
  net_unblended_cost?: number
  usage_quantity?: number
  [key: string]: number | undefined
}

/**
 * Individual cost result from costs API
 */
export interface CostResult {
  start: string
  end: string
  service: string
  usage_type: string
  values: CostValues
}

/**
 * Usage type aggregate structure
 */
export interface UsageTypeAggregate {
  sum?: number
  avg?: number
  max?: number
  min?: number
}

/**
 * Cost type aggregates nested structure:
 * service -> cost_type -> usage_type -> aggregate
 */
export interface CostTypeAggregates {
  total?: number // Service-level total
  [usageType: string]: UsageTypeAggregate | number | undefined
}

/**
 * Service aggregates nested structure:
 * service -> cost_type -> cost type aggregates
 */
export interface ServiceAggregates {
  [costType: string]: CostTypeAggregates
}

/**
 * Cost aggregates structure:
 * service -> cost_type -> usage_type -> aggregate
 */
export interface CostAggregates {
  [service: string]: ServiceAggregates
}

/**
 * Cost query information for transparency
 */
export interface CostQueryInfo {
  time_period: { start: string; end: string }
  granularity: string
  services: string[]
  regions: string[]
  group_by: string[]
  metrics: string[]
  tags?: Record<string, string[]>
  aws_cli_command: string
  console_url: string
  aggregation_note: string
}

/**
 * Costs API response structure
 */
export interface CostsApiResponse {
  metadata: ApiMetadata
  results: CostResult[]
  aggregates?: CostAggregates
  query_info?: CostQueryInfo
}

/**
 * Query parameters for costs API
 */
export interface CostsQueryParams {
  startDate?: string | Date
  endDate?: string | Date
}

